File: java/driver/src/main/java/com/google/mapsplatform/transportation/sample/driver/MainActivity.java
/* Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.mapsplatform.transportation.sample.driver;

import android.annotation.SuppressLint;
import android.app.Application;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.TextView;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;
import com.google.android.gms.maps.GoogleMap.CameraPerspective;
import com.google.android.libraries.mapsplatform.transportation.driver.api.ridesharing.RidesharingDriverApi;
import com.google.android.libraries.navigation.NavigationApi;
import com.google.android.libraries.navigation.Navigator;
import com.google.android.libraries.navigation.SupportNavigationFragment;
import com.google.android.material.snackbar.Snackbar;
import com.google.common.base.Joiner;
import com.google.common.collect.ImmutableList;
import com.google.common.util.concurrent.FutureCallback;
import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.mapsplatform.transportation.sample.driver.dialog.VehicleDialogFragment;
import com.google.mapsplatform.transportation.sample.driver.state.TripStatus;
import java.net.ConnectException;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/*
 * NOTE: BEFORE BUILDING THIS APPLICATION YOU MUST COPY THE GOOGLE NAVIGATION API
 * ANDROID LIBRARY (AAR FILE) INTO THE libs/ DIRECTORY OF THIS APP.
 */

/** Main activity for the DriverSDK test app. */
public final class MainActivity extends AppCompatActivity implements Presenter {

  private static final String TAG = "MainActivity";
  private static final int TRIP_VIEW_INITIAL_STATE = View.GONE;

  private final ExecutorService executor = Executors.newCachedThreadPool();

  private VehicleIdStore vehicleIdStore;
  private SupportNavigationFragment navFragment;
  private TextView simulationStatusText;
  private TextView tripIdText;
  private TextView matchedTripIdsText;
  private TextView textVehicleId;
  private Button actionButton;
  private CardView tripCard;
  private VehicleController vehicleController;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    vehicleIdStore = new VehicleIdStore(this);
    simulationStatusText = findViewById(R.id.simulation_status);
    actionButton = findViewById(R.id.action_button);
    Button editVehicleIdButton = findViewById(R.id.edit_button);
    editVehicleIdButton.setOnClickListener(this::onEditVehicleButtonClicked);
    tripCard = findViewById(R.id.trip_card);
    tripIdText = findViewById(R.id.trip_id_label);
    matchedTripIdsText = findViewById(R.id.matched_trip_ids_label);
    textVehicleId = findViewById(R.id.menu_vehicle_id);
    textVehicleId.setText(vehicleIdStore.readOrDefault());
    setupNavFragment();
    updateActionButton(TRIP_VIEW_INITIAL_STATE, null);
    actionButton.setOnClickListener(this::onActionButtonClicked);

    showTripId(VehicleController.NO_TRIP_ID);
    showMatchedTripIds(ImmutableList.of());

    // Ensure the screen stays on during navigation.
    getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

    initializeSDKs();

    Log.i(TAG, "Driver SDK version: " + RidesharingDriverApi.getDriverSdkVersion());
    Log.i(TAG, "Navigation SDK version: " + NavigationApi.getNavSDKVersion());
  }

  private void initializeSDKs() {
    NavigationApi.getNavigator(
        this,
        new NavigationApi.NavigatorListener() {
          @Override
          public void onNavigatorReady(Navigator navigator) {
            Application app = (Application) getApplicationContext();
            vehicleController =
                new VehicleController(getApplication(), navigator, executor, MainActivity.this);
            vehicleController.setPresenter(MainActivity.this);
            initVehicleAndPollTrip();
          }

          @Override
          public void onError(@NavigationApi.ErrorCode int errorCode) {
            logNavigationApiInitError(errorCode);
          }
        });
  }

  private void initVehicleAndPollTrip() {
    Application app = (Application) getApplicationContext();
    ListenableFuture<Boolean> future = vehicleController.initVehicleAndReporter(app);
    Futures.addCallback(
        future,
        new FutureCallback<Boolean>() {
          @Override
          public void onSuccess(Boolean result) {}

          @Override
          public void onFailure(Throwable e) {
            if (e instanceof ConnectException) {
              Snackbar.make(
                      tripCard, R.string.msg_provider_connection_error, Snackbar.LENGTH_INDEFINITE)
                  .setAction(R.string.button_retry, view -> initVehicleAndPollTrip())
                  .show();
            }
          }
        },
        executor);
  }

  private void setupNavFragment() {
    navFragment = SupportNavigationFragment.newInstance();
    getSupportFragmentManager()
        .beginTransaction()
        .add(R.id.nav_fragment_frame, navFragment, null)
        .setReorderingAllowed(true)
        .commit();
  }

  @SuppressWarnings("FutureReturnValueIgnored")
  private void onEditVehicleButtonClicked(View view) {
    VehicleDialogFragment fragment =
        VehicleDialogFragment.newInstance(
            vehicleIdStore.readOrDefault(),
            vehicleController.getVehicleSettings(),
            updatedVehicleSettings -> {
              textVehicleId.setText(updatedVehicleSettings.getVehicleId());

              vehicleController.updateVehicleSettings(
                  (Application) getApplicationContext(), updatedVehicleSettings);
            });

    fragment.show(getSupportFragmentManager(), "VehicleInfoDialog");
  }

  private void onActionButtonClicked(View view) {
    vehicleController.processNextState();
  }

  // Permissions are managed by the 'SplashScreenActivity'
  @SuppressLint("MissingPermission")
  private void updateCameraPerspective(Boolean isTilted) {
    navFragment.getMapAsync(
        googleMap ->
            googleMap.followMyLocation(
                isTilted ? CameraPerspective.TILTED : CameraPerspective.TOP_DOWN_NORTH_UP));
  }

  private void updateActionButton(Integer visibility, @Nullable @StringRes Integer resourceId) {
    if (resourceId != null) {
      actionButton.setText(resourceId);
    }

    actionButton.setVisibility(visibility);
  }

  @Override
  public void showTripId(String tripId) {
    if (!tripId.equals(VehicleController.NO_TRIP_ID)) {
      tripIdText.setText(getResources().getString(R.string.trip_id_label, tripId));
      return;
    }

    String noTripFoundText = getResources().getString(R.string.status_unknown);
    tripIdText.setText(getResources().getString(R.string.trip_id_label, noTripFoundText));
  }

  @Override
  public void showMatchedTripIds(List<String> tripIds) {
    matchedTripIdsText.setVisibility(View.VISIBLE);

    if (tripIds.isEmpty()) {
      String noTripFoundText = getResources().getString(R.string.status_unknown);

      matchedTripIdsText.setText(
          getResources().getString(R.string.matched_trip_ids_label, noTripFoundText));

      return;
    }

    String text = Joiner.on(", ").join(tripIds);

    matchedTripIdsText.setText(getResources().getString(R.string.matched_trip_ids_label, text));
  }

  @Override
  public void showTripStatus(TripStatus status) {
    int resourceId = R.string.status_idle;
    int buttonVisibility = View.VISIBLE;
    int cardVisibility = View.VISIBLE;

    // Determines whether the camera perspective should be tilted.
    boolean isCameraTilted = false;

    switch (status) {
      case UNKNOWN_TRIP_STATUS:
        buttonVisibility = TRIP_VIEW_INITIAL_STATE;
        cardVisibility = View.VISIBLE;
        simulationStatusText.setText(resourceId);
        break;

      case NEW:
        simulationStatusText.setText(R.string.status_new);
        resourceId = R.string.button_start_trip;
        break;

      case ENROUTE_TO_PICKUP:
        simulationStatusText.setText(R.string.status_enroute_to_pickup);
        resourceId = R.string.button_arrived_at_pickup;
        isCameraTilted = true;
        break;

      case ARRIVED_AT_PICKUP:
        simulationStatusText.setText(R.string.status_arrived_at_pickup);
        resourceId =
            vehicleController.isNextCurrentTripWaypointIntermediate()
                ? R.string.button_enroute_to_intermediate_stop
                : R.string.button_enroute_to_dropoff;
        isCameraTilted = true;
        break;

      case ENROUTE_TO_DROPOFF:
        simulationStatusText.setText(R.string.status_enroute_to_dropoff);
        resourceId = R.string.button_trip_complete;
        isCameraTilted = true;
        break;

      case ENROUTE_TO_INTERMEDIATE_DESTINATION:
        simulationStatusText.setText(R.string.status_enroute_to_intermediate_location);
        resourceId = R.string.button_arrived_at_intermediate_stop;
        isCameraTilted = true;
        break;

      case ARRIVED_AT_INTERMEDIATE_DESTINATION:
        simulationStatusText.setText(R.string.status_arrived_to_intermediate_location);
        resourceId =
            !vehicleController.isNextCurrentTripWaypointIntermediate()
                ? R.string.button_enroute_to_dropoff
                : R.string.button_enroute_to_intermediate_stop;
        isCameraTilted = true;
        break;

      case COMPLETE:
        resourceId = R.string.status_complete;
        buttonVisibility = View.GONE;
        simulationStatusText.setText(resourceId);
        break;

      case CANCELED:
        resourceId = R.string.status_canceled;
        buttonVisibility = View.GONE;
        simulationStatusText.setText(resourceId);
        break;
    }

    tripCard.setVisibility(cardVisibility);
    updateActionButton(buttonVisibility, resourceId);
    updateCameraPerspective(isCameraTilted);
  }

  @Override
  public void enableActionButton(boolean enabled) {
    actionButton.setEnabled(enabled);
  }

  private static void logNavigationApiInitError(int errorCode) {
    switch (errorCode) {
      case NavigationApi.ErrorCode.NOT_AUTHORIZED:
        Log.w(
            TAG,
            "Note: If this message is displayed, you may need to check that your API_KEY is"
                + " specified correctly in AndroidManifest.xml and is been enabled to"
                + " access the Navigation API.");
        break;
      case NavigationApi.ErrorCode.TERMS_NOT_ACCEPTED:
        Log.w(
            TAG,
            "Error loading Navigation API: User did not " + "accept the Navigation Terms of Use.");
        break;
      default:
        Log.w(TAG, String.format("Error loading Navigation API: %d", errorCode));
    }
  }

  @Override
  protected void onDestroy() {
    vehicleController.cleanUp();

    super.onDestroy();
  }
}

File: kotlin/kotlin-driver/src/main/kotlin/com/google/mapsplatform/transportation/sample/kotlindriver/MainActivity.kt
/* Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.mapsplatform.transportation.sample.kotlindriver

import android.annotation.SuppressLint
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.WindowManager
import android.widget.Button
import android.widget.TextView
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import androidx.cardview.widget.CardView
import androidx.lifecycle.lifecycleScope
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.GoogleMap.CameraPerspective
import com.google.android.libraries.mapsplatform.transportation.driver.api.ridesharing.RidesharingDriverApi
import com.google.android.libraries.navigation.NavigationApi
import com.google.android.libraries.navigation.NavigationApi.NavigatorListener
import com.google.android.libraries.navigation.Navigator
import com.google.android.libraries.navigation.SupportNavigationFragment
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import com.google.mapsplatform.transportation.sample.kotlindriver.dialog.VehicleDialogFragment
import com.google.mapsplatform.transportation.sample.kotlindriver.provider.ProviderUtils
import com.google.mapsplatform.transportation.sample.kotlindriver.provider.request.VehicleSettings
import com.google.mapsplatform.transportation.sample.kotlindriver.provider.service.LocalProviderService
import com.google.mapsplatform.transportation.sample.kotlindriver.state.TripStatus
import java.net.ConnectException
import java.util.concurrent.Executors
import kotlinx.coroutines.launch

/*
 * NOTE: BEFORE BUILDING THIS APPLICATION YOU MUST COPY THE GOOGLE NAVIGATION API
 * ANDROID LIBRARY (AAR FILE) INTO THE libs/ DIRECTORY OF THIS APP.
 */
/** Main activity for the DriverSDK test app. */
class MainActivity : AppCompatActivity(), Presenter {
  private val executor = Executors.newCachedThreadPool()
  private lateinit var localSettings: LocalSettings
  private lateinit var navFragment: SupportNavigationFragment
  private lateinit var simulationStatusText: TextView
  private lateinit var tripIdText: TextView
  private lateinit var matchedTripIdsText: TextView
  private lateinit var textVehicleId: TextView
  private lateinit var actionButton: Button
  private lateinit var tripCard: CardView
  private lateinit var vehicleController: VehicleController

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    localSettings = LocalSettings(this)
    simulationStatusText = findViewById(R.id.simulation_status)
    actionButton = findViewById(R.id.action_button)
    val editVehicleIdButton = findViewById<Button>(R.id.edit_button)
    editVehicleIdButton.setOnClickListener { onEditVehicleButtonClicked() }
    tripCard = findViewById(R.id.trip_card)
    tripIdText = findViewById(R.id.trip_id_label)
    matchedTripIdsText = findViewById(R.id.matched_trip_ids_label)
    textVehicleId = findViewById(R.id.menu_vehicle_id)
    textVehicleId.text = localSettings.getVehicleId()
    setupNavFragment()
    updateActionButton(TRIP_VIEW_INITIAL_STATE, null)
    actionButton.setOnClickListener { onActionButtonClicked() }
    showTripId(VehicleController.NO_TRIP_ID)
    showMatchedTripIds(emptyList())
    // Ensure the screen stays on during navigation.
    window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    initializeSDKs()
    Log.i(TAG, "Driver SDK version: " + RidesharingDriverApi.getDriverSdkVersion())
    Log.i(TAG, "Navigation SDK version: " + NavigationApi.getNavSDKVersion())
  }

  private fun initializeSDKs() {
    NavigationApi.getNavigator(
      this,
      object : NavigatorListener {
        override fun onNavigatorReady(navigator: Navigator) {
          vehicleController =
            VehicleController(
              navigator,
              executor,
              context = this@MainActivity,
              coroutineScope = this@MainActivity.lifecycleScope,
              providerService =
                LocalProviderService(
                  LocalProviderService.createRestProvider(
                    ProviderUtils.getProviderBaseUrl(application)
                  ),
                ),
              localSettings,
            )
          vehicleController.setPresenter(this@MainActivity)
          initVehicleAndPollTrip()
        }

        override fun onError(@NavigationApi.ErrorCode errorCode: Int) {
          logNavigationApiInitError(errorCode)
        }
      }
    )
  }

  private fun initVehicleAndPollTrip() {
    lifecycleScope.launch {
      try {
        vehicleController.initVehicleAndReporter(application)
      } catch (e: ConnectException) {
        Snackbar.make(
            tripCard,
            R.string.msg_provider_connection_error,
            BaseTransientBottomBar.LENGTH_INDEFINITE
          )
          .setAction(R.string.button_retry) { initVehicleAndPollTrip() }
          .show()
      }
    }
  }

  private fun setupNavFragment() {
    navFragment = SupportNavigationFragment.newInstance()
    supportFragmentManager
      .beginTransaction()
      .add(R.id.nav_fragment_frame, navFragment, null)
      .setReorderingAllowed(true)
      .commit()
  }

  private fun onEditVehicleButtonClicked() {
    val fragment: VehicleDialogFragment =
      VehicleDialogFragment.newInstance(
        localSettings,
        vehicleController.vehicleSettings,
        object : VehicleDialogFragment.OnDialogResultListener {
          override fun onResult(vehicleSettings: VehicleSettings) {
            textVehicleId.text = vehicleSettings.vehicleId
            localSettings.saveVehicleId(vehicleSettings.vehicleId)

            lifecycleScope.launch {
              vehicleController.updateVehicleSettings(application, vehicleSettings)
            }
          }
        }
      )
    fragment.show(supportFragmentManager, "VehicleInfoDialog")
  }

  private fun onActionButtonClicked() {
    vehicleController.processNextState()
  }

  // Permissions are managed by the 'SplashScreenActivity'
  @SuppressLint("MissingPermission")
  private fun updateCameraPerspective(isTilted: Boolean) {
    navFragment.getMapAsync { googleMap: GoogleMap ->
      googleMap.followMyLocation(
        if (isTilted) CameraPerspective.TILTED else CameraPerspective.TOP_DOWN_NORTH_UP
      )
    }
  }

  private fun updateActionButton(visibility: Int, @StringRes resourceId: Int?) {
    resourceId?.also { actionButton.setText(it) }
    actionButton.visibility = visibility
  }

  override fun showTripId(tripId: String) {
    if (tripId != VehicleController.NO_TRIP_ID) {
      tripIdText.text = resources.getString(R.string.trip_id_label, tripId)
      return
    }
    val noTripFoundText = resources.getString(R.string.status_unknown)
    tripIdText.text = resources.getString(R.string.trip_id_label, noTripFoundText)
  }

  override fun showMatchedTripIds(tripIds: List<String>) {
    val text =
      if (tripIds.isEmpty()) resources.getString(R.string.status_unknown)
      else tripIds.joinToString()

    matchedTripIdsText.text = resources.getString(R.string.matched_trip_ids_label, text)
  }

  override fun showTripStatus(status: TripStatus) {
    var resourceId = R.string.status_idle
    var buttonVisibility = View.VISIBLE
    var cardVisibility = View.VISIBLE

    // Determines whether the camera perspective should be tilted.
    var isCameraTilted = false
    when (status) {
      TripStatus.UNKNOWN_TRIP_STATUS -> {
        buttonVisibility = TRIP_VIEW_INITIAL_STATE
        cardVisibility = View.VISIBLE
        simulationStatusText.setText(resourceId)
      }
      TripStatus.NEW -> {
        simulationStatusText.setText(R.string.status_new)
        resourceId = R.string.button_start_trip
      }
      TripStatus.ENROUTE_TO_PICKUP -> {
        simulationStatusText.setText(R.string.status_enroute_to_pickup)
        resourceId = R.string.button_arrived_at_pickup
        isCameraTilted = true
      }
      TripStatus.ARRIVED_AT_PICKUP -> {
        simulationStatusText.setText(R.string.status_arrived_at_pickup)
        resourceId =
          if (vehicleController.isNextCurrentTripWaypointIntermediate()) {
            R.string.button_enroute_to_intermediate_stop
          } else R.string.button_enroute_to_dropoff
        isCameraTilted = true
      }
      TripStatus.ENROUTE_TO_DROPOFF -> {
        simulationStatusText.setText(R.string.status_enroute_to_dropoff)
        resourceId = R.string.button_trip_complete
        isCameraTilted = true
      }
      TripStatus.ENROUTE_TO_INTERMEDIATE_DESTINATION -> {
        simulationStatusText.setText(R.string.status_enroute_to_intermediate_location)
        resourceId = R.string.button_arrived_at_intermediate_stop
        isCameraTilted = true
      }
      TripStatus.ARRIVED_AT_INTERMEDIATE_DESTINATION -> {
        simulationStatusText.setText(R.string.status_arrived_to_intermediate_location)
        resourceId =
          if (vehicleController.isNextCurrentTripWaypointIntermediate()) {
            R.string.button_enroute_to_intermediate_stop
          } else R.string.button_enroute_to_dropoff
        isCameraTilted = true
      }
      TripStatus.COMPLETE -> {
        resourceId = R.string.status_complete
        buttonVisibility = View.GONE
        simulationStatusText.setText(resourceId)
      }
      TripStatus.CANCELED -> {
        resourceId = R.string.status_canceled
        buttonVisibility = View.GONE
        simulationStatusText.setText(resourceId)
      }
    }
    tripCard.visibility = cardVisibility
    updateActionButton(buttonVisibility, resourceId)
    updateCameraPerspective(isCameraTilted)
  }

  override fun enableActionButton(enabled: Boolean) {
    actionButton.isEnabled = enabled
  }

  override fun onDestroy() {
    vehicleController.cleanUp()
    super.onDestroy()
  }

  companion object {
    private const val TAG = "MainActivity"
    private const val TRIP_VIEW_INITIAL_STATE = View.GONE
    private fun logNavigationApiInitError(errorCode: Int) {
      when (errorCode) {
        NavigationApi.ErrorCode.NOT_AUTHORIZED ->
          Log.w(
            TAG,
            "Note: If this message is displayed, you may need to check that your API_KEY is" +
              " specified correctly in AndroidManifest.xml and is been enabled to" +
              " access the Navigation API."
          )
        NavigationApi.ErrorCode.TERMS_NOT_ACCEPTED ->
          Log.w(
            TAG,
            "Error loading Navigation API: User did not accept the Navigation Terms of Use."
          )
        else -> Log.w(TAG, "Error loading Navigation API: $errorCode")
      }
    }
  }
}
